crates/noirc_evaluator/src/ssa/ssa_gen/mod.rs

```diff
@@ -269,6 +302,38 @@ impl<'a> FunctionContext<'a> {
         Self::map_type(element_type, |typ| {
             let offset = self.make_offset(base_index, field_index);
             field_index += 1;
+
+            let array_type = &self.builder.type_of_value(array);
+            match array_type {
+                // Prepare a slice access
+                // Check that the index being used is less than the dynamic slice length
+                Type::Slice(_) => {
+                    let array_len =
+                        max_length.expect("ICE: a length must be supplied for indexing slices");
+                    // Check the type of the index value for valid comparisons
+                    let array_len = match self.builder.type_of_value(index) {
+                        Type::Numeric(numeric_type) => match numeric_type {
+                            // If the index itself is an integer, keep the array length as a Field
+                            NumericType::Unsigned { .. } | NumericType::Signed { .. } => array_len,
+                            // If the index and the array length are both Fields we will not be able to perform a less than comparison on them.
+                            // Thus, we cast the array length to a u64 before performing the less than comparison
+                            NumericType::NativeField => self.builder.insert_cast(
+                                array_len,
+                                Type::Numeric(NumericType::Unsigned { bit_size: 64 }),
+                            ),
+                        },
+                        _ => unreachable!("ICE: array index must be a numeric type"),
+                    };
+
+                    let is_offset_out_of_bounds =
+                        self.builder.insert_binary(index, BinaryOp::Lt, array_len);
+                    self.builder.insert_constrain(is_offset_out_of_bounds);
```

Comment:
Why do we need to insert automatic `assert(i < slice.len())` calls for slices again? I was under the impression we already checked this within array_get and array_set.

Reply:
Brillig is not checking it :thinking:  maybe it doesn't need to be checked in SSA if we add the check to brillig gen?

Reply:
In order to merge slices during flattening, the smaller slice to be merged is filled with placeholder data. So the `Value::Array` stored at a Slice type can have extra data, and this is why I added the test in `compile_failure`. Without this check the `slice_access_failure` test would pass execution. The check within array_get and array_set simply checks the length of the list stored in the SSA value. This was accurate for slices before merging.

Reply:
> Brillig is not checking it 🤔 maybe it doesn't need to be checked in SSA if we add the check to brillig gen?

Check in array get and array set?